src/rdkafka_telemetry.c

```diff
@@ -0,0 +1,679 @@
+ */
+static int
+rd_kafka_filter_broker_by_GetTelemetrySubscription(rd_kafka_broker_t *rkb,
+                                                   void *opaque) {
+        int features;
+        if (rd_kafka_broker_ApiVersion_supported0(
+                rkb, RD_KAFKAP_GetTelemetrySubscriptions, 0, 0, &features,
+                rd_false) != -1)
+                return 0;
+        return 1;
+}
+
+/**
+ * @brief Cleans up the rk.rk_telemetry struct and frees any allocations.
+ *
+ * @param clear_control_flow_fields This determines if the control flow fields
+ *                                  need to be cleared. This should only be set
+ *                                  to true if the rk is terminating.
+ * @locality main thread
+ * @locks none
+ * @locks_acquired rk_telemetry.lock
+ */
+void rd_kafka_telemetry_clear(rd_kafka_t *rk,
+                              rd_bool_t clear_control_flow_fields) {
+        rd_kafka_broker_t *rkb;
+        if (clear_control_flow_fields) {
+                mtx_lock(&rk->rk_telemetry.lock);
+                if (rk->rk_telemetry.preferred_broker) {
+                        rd_kafka_broker_destroy(
+                            rk->rk_telemetry.preferred_broker);
+                        rk->rk_telemetry.preferred_broker = NULL;
+                }
+                mtx_unlock(&rk->rk_telemetry.lock);
+                mtx_destroy(&rk->rk_telemetry.lock);
+                cnd_destroy(&rk->rk_telemetry.termination_cnd);
+        }
+
+        if (rk->rk_telemetry.accepted_compression_types_cnt) {
+                rd_free(rk->rk_telemetry.accepted_compression_types);
+                rk->rk_telemetry.accepted_compression_types     = NULL;
+                rk->rk_telemetry.accepted_compression_types_cnt = 0;
+        }
+
+        if (rk->rk_telemetry.requested_metrics_cnt) {
+                size_t i;
+                for (i = 0; i < rk->rk_telemetry.requested_metrics_cnt; i++)
+                        rd_free(rk->rk_telemetry.requested_metrics[i]);
+                rd_free(rk->rk_telemetry.requested_metrics);
+                rd_free(rk->rk_telemetry.matched_metrics);
+                rk->rk_telemetry.requested_metrics     = NULL;
+                rk->rk_telemetry.requested_metrics_cnt = 0;
+                rk->rk_telemetry.matched_metrics       = NULL;
+                rk->rk_telemetry.matched_metrics_cnt   = 0;
+        }
+        rk->rk_telemetry.telemetry_max_bytes = 0;
+}
+
+/**
+ * @brief Sets the telemetry state to TERMINATED and signals the conditional
+ * variable
+ *
+ * @locality main thread
+ * @locks none
+ * @locks_acquired rk_telemetry.lock
+ */
+static void rd_kafka_telemetry_set_terminated(rd_kafka_t *rk) {
+        rd_dassert(thrd_is_current(rk->rk_thread));
+
+        rd_kafka_dbg(rk, TELEMETRY, "TERM",
+                     "Setting state to TERMINATED and signalling");
+
+        rk->rk_telemetry.state = RD_KAFKA_TELEMETRY_TERMINATED;
+        mtx_lock(&rk->rk_telemetry.lock);
```

Comment:
Here it should stop the timer, it happened that it reached a dassert because it was scheduled, only with devel mode enabled, but it's annoying for testing.
```suggestion
        rd_kafka_timer_stop(&rk->rk_timers, &rk->rk_telemetry.request_timer,
                            1 /*lock*/);
        mtx_lock(&rk->rk_telemetry.lock);
```

Reply:
Updated